test/components/structures/MatrixChat-test.tsx

```diff
@@ -388,6 +401,117 @@ describe("<MatrixChat />", () => {
                     });
                 });
             });
+
+            describe("logout", () => {
+                let logoutClient!: ReturnType<typeof getMockClientWithEventEmitter>;
+                const call1 = { disconnect: jest.fn() } as unknown as Call;
+                const call2 = { disconnect: jest.fn() } as unknown as Call;
+
+                const dispatchLogoutAndWait = async (): Promise<void> => {
+                    defaultDispatcher.dispatch({
+                        action: "logout",
+                    });
+
+                    await flushPromises();
+                };
+
+                beforeEach(() => {
+                    // stub out various cleanup functions
+                    jest.spyOn(LegacyCallHandler.instance, "hangupAllCalls")
+                        .mockClear()
+                        .mockImplementation(() => {});
+                    jest.spyOn(voiceBroadcastUtils, "cleanUpBroadcasts").mockImplementation(async () => {});
+                    jest.spyOn(PosthogAnalytics.instance, "logout").mockImplementation(() => {});
+                    jest.spyOn(EventIndexPeg, "deleteEventIndex").mockImplementation(async () => {});
+
+                    jest.spyOn(CallStore.instance, "activeCalls", "get").mockReturnValue(new Set([call1, call2]));
+
+                    mockPlatformPeg({
+                        destroyPickleKey: jest.fn(),
+                    });
+
+                    logoutClient = getMockClientWithEventEmitter(getMockClientMethods());
+                    mockClient = getMockClientWithEventEmitter(getMockClientMethods());
+                    mockClient.logout.mockResolvedValue({});
+                    mockClient.getDeviceId.mockReturnValue(deviceId);
+                    // this is used to create a temporary client to cleanup after logout
+                    jest.spyOn(MatrixJs, "createClient").mockClear().mockReturnValue(logoutClient);
+
+                    jest.spyOn(logger, "warn").mockClear();
```

Comment:
why do we need all these `mockClear`s *before* running tests? Surely that should happen *after* the test that sets up the mock, to keep tests independent?

Reply:
`mockClear` just resets the mock call counts
Technically yes the `mockClear` part could/should be in the `afterEach`, but I think it makes more litter in the test setup to add all the spies/mocks in `beforeEach` and then clear them in `afterEach`
If you're passionate about it I can split them up.

Reply:
I'm just worried that there is a risk of effects from one test leaking into another test and producing unexpected results.

What's to stop the next test (which might be in a completely different file) from getting the mocked version of `createClient` instead of the real thing?

Reply:
jest spies are scoped per test file

Reply:
added a `restoreAllMocks` to the top level afterEach

Reply:
> jest spies are scoped per test file

Ah, I hadn't quite grokked that. I couldn't find it in the jest docs.